moveit_experimental/moveit_jog_arm/test/jog_cpp_interface_test.cpp

```diff
@@ -45,57 +45,152 @@
 // Testing
 #include <gtest/gtest.h>
 
-// Main class
-#include <moveit_jog_arm/jog_cpp_interface.h>
+// Jog Arm
+#include <moveit_jog_arm/jog_arm.h>
+#include <moveit_jog_arm/make_shared_from_pool.h>
 
 static const std::string LOGNAME = "jog_cpp_interface_test";
 
 namespace moveit_jog_arm
 {
-class TestJogCppInterface : public ::testing::Test
+class JogArmFixture : public ::testing::Test
 {
 public:
   void SetUp() override
   {
-    nh_.reset(new ros::NodeHandle("~"));
-
     // Load the planning scene monitor
     planning_scene_monitor_ = std::make_shared<planning_scene_monitor::PlanningSceneMonitor>("robot_description");
     planning_scene_monitor_->startSceneMonitor();
     planning_scene_monitor_->startStateMonitor();
+    planning_scene_monitor_->startWorldGeometryMonitor(
+        planning_scene_monitor::PlanningSceneMonitor::DEFAULT_COLLISION_OBJECT_TOPIC,
+        planning_scene_monitor::PlanningSceneMonitor::DEFAULT_PLANNING_SCENE_WORLD_TOPIC,
+        false /* skip octomap monitor */);
+
+    // Create jog_arm
+    jog_arm_ = std::make_shared<JogArm>(nh_, planning_scene_monitor_);
   }
   void TearDown() override
   {
   }
 
+  bool waitForFirstStatus()
+  {
+    auto msg =
+        ros::topic::waitForMessage<std_msgs::Int8>(jog_arm_->getParameters().status_topic, nh_, ros::Duration(15));
+    return static_cast<bool>(msg);
+  }
+
 protected:
-  std::unique_ptr<ros::NodeHandle> nh_;
+  ros::NodeHandle nh_;
   planning_scene_monitor::PlanningSceneMonitorPtr planning_scene_monitor_;
-};  // class TestJogCppInterface
+  moveit_jog_arm::JogArmPtr jog_arm_;
+};  // class JogArmFixture
 
-TEST_F(TestJogCppInterface, InitTest)
+TEST_F(JogArmFixture, StartStopTest)
 {
-  moveit_jog_arm::JogCppInterface jog_cpp_interface(planning_scene_monitor_);
-  ros::Duration(1).sleep();  // Give the started thread some time to run
+  jog_arm_->start();
+  EXPECT_TRUE(waitForFirstStatus()) << "Timeout waiting for Status message";
+  jog_arm_->stop();
+
+  ros::Duration(1.0).sleep();
+
+  // Start and stop again
+  jog_arm_->start();
+  EXPECT_TRUE(waitForFirstStatus()) << "Timeout waiting for Status message";
+  jog_arm_->stop();
 }
 
-// TODO(davetcoleman): due to many blocking checks for ROS messages, and
-// an abundance of threads, unit tests are not currently feasible for the
-// cpp interface. This should be addressed in future re-factors
+TEST_F(JogArmFixture, SendTwistStampedTest)
+{
+  jog_arm_->start();
+  EXPECT_TRUE(waitForFirstStatus()) << "Timeout waiting for Status message";
+
+  auto parameters = jog_arm_->getParameters();
+
+  // count trajectory messages sent by jog_arm
+  size_t received_count = 0;
+  boost::function<void(const trajectory_msgs::JointTrajectoryConstPtr&)> traj_callback =
```

Comment:
use `auto` for the type here?

Reply:
I can't or it'll make a std::function and not a boost::function and ROS won't accept that.